Add usePreparedEffect hook and withPreparedEffect higher order component

eikhr opened the pull request:
I have added two new functions intended to replace the `prepared` and `dispatched` functions, with an API that more closely matches the way we use the library in lego-webapp. I intend to remove prepared and dispatched, but I figured it was better to do that in a separate update.

The new API is created to mimic how `useEffect` works, except that the effect will be run on the server the first time instead of the client when server-side rendering. In addition i have added a higher order component that wraps the new hook, to make the transition in `lego-webapp` easier.

# API

## `usePreparedEffect()`
It works exactly like a normal `useEffect()` except it runs during `prepare()` instead of on the client when SSR is used.

You might use it like this:
```js
const dispatch = useDispatch();
usePreparedEffect('fetchItems', () => {
  dispatch(fetchAll({ query: props.location.search }));
}, [dispatch, props.location.search]);
```

It also has an option `runSync` which works like `awaitOnSsr` in `prepared`, meaning it awaits the effect before preparing children.

This might be useful in the future, but for the time beeing our routes are basically just compositions of higher order components, which is why i created the `withPreparedEffect` HOC.
## `withPreparedEffect()`
This is basically just a wrapper for adding a `usePreparedEffect()` hook. Just like `usePreparedEffect` it takes an effect function as an argument, and it will get passed all the component's props. It also takes a function that creates the dependency array, which also gets passed the props.

It might be used like this:
```js
export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withPreparedEffect(
    'fetchItems',
    (props) => {
      props.dispatch(fetchAll({ query: props.location.search }));
    },
    (props) => [props.dispatch, props.location.search],
  )
)(Component)
```

eikhr committed:
Add usePreparedEffect hook

eikhr committed:
Support awaiting side effect before traversing children in useEffect

eikhr committed:
fix bug where parallel "prepare"s would await wrong promises

eikhr committed:
Add withPreparedEffect function

eikhr committed:
Support error handling in usePreparedEffect

eikhr committed:
Update README with new functions

danielyanghansen commented on src/prepare.ts:
```diff
@@ -150,12 +156,16 @@ async function prepareElement(
-      return [functionElement.type(functionElement.props), context];
+      setDispatcherContext(dispatcher, context);
+      registerDispatcher(dispatcher);
+      const children = functionElement.type(functionElement.props);
+      await Promise.all(popAwaitImmediatelyPromises(dispatcher));
+      return [children, context];
```
Is `children` more than one `ReactNode` here?
Considering the function output signature of `Promise<[ReactNode, PrepareContext, Promise<void>?]> `

eikhr replied:
A `ReactNode` could be either a single react node or an array of nodes. This is why we use `React.Children.toArray` below to make sure that we are dealing with an array of elements.
https://reactnative.dev/docs/react-node

danielyanghansen commented on src/prepare.ts:
```diff
@@ -150,12 +156,16 @@ async function prepareElement(
-      return [functionElement.type(functionElement.props), context];
+      setDispatcherContext(dispatcher, context);
+      registerDispatcher(dispatcher);
+      const children = functionElement.type(functionElement.props);
```
```suggestion
      const children : ReactNode = functionElement.type(functionElement.props);
```

danielyanghansen commented on src/usePreparedEffect.ts:
```diff
@@ -0,0 +1,38 @@
+  prepareFunction: PrepareHookFunction,
+  deps?: DependencyList,
+  opts: PreparedEffectOptions = {},
+) => {
```
You should declare the function output signature here.
I'm assuming that both functions called on internally here:
 - `prepareFunction()`
 - `useEffect()`
... have return types of `(void)`, but if they don't, you should probably create a discriminated union type for what is an allowed output from this function

eikhr replied:
Added:)

danielyanghansen commented on src/usePreparedEffect.ts:
```diff
@@ -0,0 +1,38 @@
+import { DependencyList, useEffect, useRef } from 'react';
+import { __REACT_PREPARE__ } from './constants';
+
+export interface PreparedEffectOptions {
+  runSync?: boolean;
+}
```
You mentioned a couple of options in the `README`:
 - `awaitImmediately`
 - `runOnClient`

Should those also be part of this interface, or am I misunderstanding how these options are passed down?

eikhr replied:
They are removed now:)

danielyanghansen commented on src/utils/dispatcher.ts:
```diff
@@ -57,17 +96,43 @@ const dispatcher: Dispatcher = {
 
-export const dispatcherIsRegistered = (): boolean =>
-  ReactInternals.ReactCurrentDispatcher.current === dispatcher;
+export const popPreparedHookPromises = (
+  dispatcher: Dispatcher,
+): Promise<unknown>[] => {
+  const promises = dispatcher[__REACT_PREPARE__].hookPromises;
+  dispatcher[__REACT_PREPARE__].hookPromises = [];
+  return promises;
```
Is it always the case that we want to pop all of the hook promises at once?
You'll know more about this than me, but it's nice to confirm if this is intentional

danielyanghansen replied:
Same for the other pop functions below

eikhr replied:
Yes, we basically add all `usePreparedEffect`-promises we find while parsing the component tree to the dispatcher. When the time comes to await the promises, we then pop all of them and await them at the same time using `Promise.all`. 

Normally this will happen at the end of `prepare`, after parsing the entire component-tree.

In the case of `usePreparedEffect`s with `runSync=true`, this will instead happen right away after parsing the component. The reason we need to pop an array of promises here is the case where one has multiple `runSync=true`-hooks in a single component, in which case we await these hooks simultainiously before parsing the children. (Ideally we would await the first one before parsing the second, but AFAIK this is not possible)

danielyanghansen commented on src/tests/prepare.spec.jsx:
```diff
@@ -159,6 +160,61 @@ describe('prepare', () => {
+      );
+    } catch (err) {
+      assert.equal(err.message, 'Err', 'Should throw the correct error');
+      assert(doAsyncSideEffect.calledOnce, 'Should be called once times');
```
```suggestion
      assert(doAsyncSideEffect.calledOnce, 'Should be called one time');
```

danielyanghansen commented on src/tests/prepare.spec.jsx:
```diff
@@ -695,4 +751,28 @@ describe('prepare', () => {
+    prepare(<SlowComponent />);
+    await prepare(<FastComponent />);
+
+    assert.equal(numAwaited, 1, 'Only fast effect has been awaited');
```
Might be a point to also assert that the slow effect has been awaited after the full second

eikhr replied:
Nah, it would make the test too slow. And we have other tests that test that effects are awaited and run.

eikhr committed:
Bump version

danielyanghansen commented on src/tests/usePreparedEffect.spec.jsx:
```diff
@@ -0,0 +1,280 @@
+    eval(prepareClientCode);
+    render(<OtherComponent />);
+
+    assert(
+      prepareFunction.calledOnce,
+      'prepareFunction has been called exactly once',
+    );
```
Superficial assertion?

eikhr replied:
true

danielyanghansen commented on src/tests/usePreparedEffect.spec.jsx:
```diff
@@ -0,0 +1,280 @@
+    );
+    assert(
+      otherEffectFunction.calledOnce,
+      'prepareFunction has been called exactly once',
```
```suggestion
      'otherEffectFunction has been called exactly once',
```

danielyanghansen reviewed (changes requested):
A few small nitpicks (and questions), but it looks pretty good!

LudvigHz commented on README.md:
```diff
@@ -73,10 +83,22 @@ render(
+
+`opts` is an optional configuration object:
+
+- `opts.runSync` (default: `false`): On the server, `prepare` will await the `sideEffect`-promise before traversing further down the tree. When `false` the promise will be awaited at the same time as all other prepare-promises before `prepare` returns.
```
This last part is a bit confusing tbh, maybe just write that it will be awaited before `prepare` returns, or "at the same time as all other `runSync` prepare-promises". Since the "at the same time as all other prepare-promises" here could be interpreted a bit differently.

LudvigHz commented on README.md:
```diff
@@ -73,10 +83,22 @@ render(
-
 ### API
 
+#### `usePreparedEffect(identifier: string, sideEffect: async () => Promise<void>, deps, opts)`
```
Also please add a short explanation of `deps` here as well.

eikhr commented:
Thank you very much for a thougrough review @danielyanghansen, I believe I have addressed all your comments now:) 

I have also fixed a problem @LudvigHz made me aware of, where there would be unexpected behaivour when unmounting components and later re-mounting them (think going to a different page, and later going back). This would make the `usePreparedEffect`-effect re-run only if the effect was not originally prepared on the server. It will now always re-run after re-mounting.

danielyanghansen reviewed (commented):
:shipit: Looks good!

LudvigHz commented on src/usePreparedEffect.ts:
```diff
@@ -0,0 +1,57 @@
+  { runSync = false }: PreparedEffectOptions = {},
+): void => {
+  // keep track of whether it is the initial effect-run or not, as only the first run is affected by server-preparing
+  const isInitialRunOnClient = useRef(true);
+
+  const effect: EffectCallback = () => {
+    const preparedEffects = window[__REACT_PREPARE__]?.preparedEffects;
+
+    const didRunOnServer = preparedEffects?.includes(identifier);
+
+    // If this is a re-run of the effect on the client, the dependency array must have changed, and we want to run the prepare function again.
+    // We also want to run the prepare function on the client if it was not prepared on the server.
+    const shouldRunEffect = !isInitialRunOnClient.current || !didRunOnServer;
+
+    if (shouldRunEffect) {
+      prepareFunction();
+    }
+
+    isInitialRunOnClient.current = false;
+
+    // remove the effect from the list of prepared effects on cleanup, such that it will be re-run if the component re-mounts.
+    return () => {
+      if (
+        window[__REACT_PREPARE__] &&
+        window[__REACT_PREPARE__].preparedEffects.includes(identifier)
+      ) {
+        window[__REACT_PREPARE__].preparedEffects.splice(
+          window[__REACT_PREPARE__].preparedEffects.indexOf(identifier),
+          1,
+        );
+      }
+    };
+  };
```
Note that this will make the hook re-run on every mount as well (except for hydration). But I _think_ that's what we agreed on doing?

Please do remind me though :sweat_smile:

eikhr replied:
Yes, I believe that is what we landed on there. I think it makes sense to refetch data when you go back to a page you have been on previously, just in case something has changed.

LudvigHz reviewed (approved):
This looks good to me now.

There were some pending comments here from earlier that I forgot to submit, but I think they're still relevant.

I think we are at the point where we should just test this in prod and see if it works out or not.
The small decisions that are arbitrary could be changed then based on how it works.

eikhr committed:
Fix file name for module export in package.json
